# 가설검정
- section 01 : 가설검정의 이해
- section 02 : 단일 표본 검정
- section 03 : 대응 표본 검정
- section 04 : 독립 표본 검정

## section 01 : 가설 검정의 이해

### 귀무가설과 대립가설
- 통계적 가설검정은 두 가지 가설을 설정한다.
- 귀무가설(H0) : 기존에 알려진 사실
    - 아메리카노 한 잔의 원두는 20g이다.
- 대립가설(H1) : 밝히려고 하는 가설
    - 아메리카노 한 잔의 원두는 20g이 아니다.

### 가설검성의 결론
- 귀무가설과 대립가설이 존재할 때 가설검정의 결과는 귀무가설 or 대립가설을 채택하게 됨.
- 귀무가설이 옳을 경우 -> 귀무가설을 기각하지 못한다.
- 대립가설이 옳을 경우 -> 귀무가설을 기각하고 대립가설을 채택한다.

### 가설검정 프로세스
- 가설 검정에 기각역을 이용하는 방법과 유의확률을 이용하는 방법이 있다.

1. 통계적 가설(귀무 가설과 대립가설)
    - 검증하고자 하는 가설을 설정한다. 귀무가설은 기존에 알려진 일반적인 주장이고, 대립가설은 입증하고자 하는 가설이다.

2. 유의 수준 결정
    - 유의수준은 귀무가설을 기각하기 위해 필요한 기준값이다.
    - 일반적으로 0.05또는 0.01과 같은 값을 사용한다.
3. 검정 통계량 및 유의확률 계산
    - 표본 데이터의 검정 통계량을 계산한다.
    - 검정 통계량 가설을 평가하기 위해 사용되는 값으로 검정 통계량을 사용해 유의확률을 계산한다.
4. 결과 도출
    - 유의확률이 유의수준보다 작으면 귀무가설을 기각
    - 유의확률이 유의수준보다 크면 귀무가설을 채택
    

## section 02 : 단일 표본 검정


### 단일 표본 검정
- 단일 표본 검정(One sample t-test)은 어떤 집단의 평균이 특정 값과 유의미하게 다른지를 검정하는 통계방법이다.
- 영화관에서 판매하는 팝콘이 무게는 120g으로 알려져 있다.
- 판매할 때마다 무게가 다른 것 같아 평균적으로 무게가 120g인지를 확인하고자 한다.
- 이때 단일 표본 검정을 사용할 수 있다.

### 표본
- 판매할 때 마다 무게를 모두 측정해 기록하는 것은 비효율적이므로 손님들이 팝콘을 구매할 때 무작위로 선택해 무게를 측정
- 이렇게 선택된 팝콘이 표본이다.


### t-test(t-검정)
- 표본의 크기, 표본의 평균, 표본의 표준편차 등을 활용해 기존에 알려진 120g과 다른지 t-test를 수행한다.
- t-test 결과는 p-value가 주어지는데 이 값이 얼마나 작은지에 따라 평균 120g과 같은지, 다른지를 판단할 수 있다.
- 유의 수준(0.05)보다 작으면 통계적으로 유의미하게 다르다고 결론짓는다.
- 단일 표본 t-test는 주어진 표본 데이터가 어떤 특정 값과 다르다는 것을 통계적으로 판단하는데 사용한다.



-  단일 표본 검정에서는 scipy.stats의 ttest_lsamp()를 활용한다.
- ttest_lsamp(a,popnmean,alternative)
    - a : 모집단에서 뽑은 표본 데이터
    - popmean : 비교하려는 모집단의 평균 또는 기대값
    - alternative : 대립가설 정의
        - μ > μ0 : grater, 주어진 특정 값보다 표본의 평균이 크다
        - μ < μ0 : less,   주어진 특정 값보다 표본의 평균이 작다
        - μ != μ0: two-signed, 주어진 특정 값과 표본의 평균이 같지 않다.
        

In [1]:
# 영화관에서 판매하는 팝콘 라지 사이즈의 무게는 일반적으로 120g으로 알려져 있다.
# 이제 이 데이터를 갖고 t-test를 수행한다

import pandas as pd

df = pd.DataFrame({
    'weights':[122,121,120,119,125,115,121,118,117,127,
               123,129,119,124,114,126,122,124,121,116,
               120,123,127,118,122,117,124,125,123,2532]
})

### 통계적 가설 설정
- 귀무가설은 '기존에 알려진 사실' 또는 '변화가 없음' 또는 '차이가 없음'을 주장하는 가설이다.
- 이 가설검정에서 귀무가설은 '팝콘 라지 사이즈의 평균 무게는 120g'
- 대립가설은 귀무가설과 반대의 주장을 하는 가설로 '팝콘 라지 사이즈의 평균 무게는 120g

- 귀무가설 : μ = μ0 -> 팝콘 라지 사이즈의 평균 무게는 120g이다.
- 대립가설 : μ != μ0 -> 팝콘 라지 사이즈의 평균 무게는 120g이 아니다.

### 유의수준 결정
- 0.05 -> 시험문제에서는 기준을 안내

### 검정 통계량과 유의확률
- 단일 표본 검정을 수행하는 ttest-1samp(표본데이터,알려진 평균값)함수를 통해 검정 통계량과 유의확률(p-value)를 반환받을 수 있음.

### 결과 도출
- p-value가 유의수준 0.05보다 작다면 팝콘 라지 사이즈의 평균 무게가 120g이 아니라는 증거
    - 귀무가설 기각 -> 대립가설 채택
- p-value가 유의수준 0.05보다 크다면 팝콘 라지 사이즈의 평균 무게가 120g이라는 기존 가설을 기각할 충분한 증거를 얻지 못함.
    - 귀무가설 채택하고 -> 대립가설 기각
    

In [3]:
from scipy import stats

t_statistic, p_value = stats.ttest_1samp(df['weights'],120)
print('t_statistic : ',t_statistic)
print('p_value : ',p_value)

t_statistic :  1.0179880333213098
p_value :  0.3171017160932224


In [6]:
stats.ttest_1samp(df['weights'],120)

TtestResult(statistic=np.float64(1.0179880333213098), pvalue=np.float64(0.3171017160932224), df=np.int64(29))

### 양측 검정과 단측 검정

#### 1. 양측 검정(Two-tailed test)
- 귀무가설, 기존에 알려진 사실보다 크거나 작음에 대해 모두 검증
- 대립가설을 팝콘 라지 사이즈의 평균 무게는 120g이 아니다. 라고 설정한다면 양측 검정이다.

#### 2. 단측 검정(One-tailed test)
- 귀무가설, 기존에 알려진 사실보다 크거나 작음에 대해 검정한다.
- 팝콘 라지 사이즈의 평균 무게는 120g 보다 크다.
    - 평균 무게인 120g보다 클 경우만 고려함.

기본적으로 ttest_1samp()함수는 양측 검정을 수행
단측 검정이 필요하면 alternative 파라미터 활용


#### 양측 검정
- tttest_1samp() 함수의 기본값은 alternative='two-sided'다 
- 귀무가설 : 팝콘 라지 사이즈의 평균 무게는 120g이다.
- 대립가설 : 팝콘 라지 사이즈의 평균 무게는 120g이 아니다.
- 가설 검정 결론 : 유의수준 0.05하에서 p-value가 0.04이므로 귀무가설 기각, 대립가설 채택


In [9]:
print(stats.ttest_1samp(df['weights'],120, alternative='two-sided'))


TtestResult(statistic=np.float64(1.0179880333213098), pvalue=np.float64(0.3171017160932224), df=np.int64(29))


#### 단측 검정
- 귀무가설 : 팝콘 라지 사이즈의 평균 무게는 120g이다.
- 대립가설 : 팝콘 라지 사이즈의 평균 무게는 120g보다 크다.


In [12]:
stats.ttest_1samp(df['weights'],120, alternative='greater')

TtestResult(statistic=np.float64(1.0179880333213098), pvalue=np.float64(0.1585508580466112), df=np.int64(29))

#### 단측 검정
- 귀무가설 : 팝콘 라지 사이즈의 평균 무게는 120g이다.
- 대립가설 : 팝콘 라지 사이즈의 평균 무게는 120g보다 작다.


In [13]:
stats.ttest_1samp(df['weights'],120, alternative='less')

TtestResult(statistic=np.float64(1.0179880333213098), pvalue=np.float64(0.8414491419533888), df=np.int64(29))

## section 03: 대응 표본 검정

### 1. 대응(쌍체) 표본 검정
- 대응 표본 검정은 동일한 그룹에 대한 시간차를 두고 두 번의 측정 결과를 비교할 때 사용
- 하나의 집단에 대한 사전,사후를 비교하며 같은 사람을 대상으로 검정한다.

```python
ttest_rel(a,b,alternative)
``` 
- a : 첫 번째 모집단에서 뽑은 표본 데이터
- b : 두 번째 모집단에서 뽑은 표본 데이터(a,b는 수는 같음)
- alternative : 대립가설 정의 : μ = (a-b)의 평균일 때
    - μ > 0 : greater , a의 평균이 b의 평균보다 크다.
    - μ < 0 : less , a의 평균이 b의 평균보다 작다.
    - μ != 0 : two-sided

In [14]:
# 문제 : 어떤 기업이 새로운 교육 프로그램을 도입해 직원의 성과를 개선하려고 한다.
# 이때 동일한 직원 그룹에 대해 교육 전과 후의 성과를 비교했다.
# 새로운 교육 프로그램이 효과가 있는지 가설검성하시오.

import pandas as pd

df = pd.DataFrame({
    'before':[85,90,92,88,86,89,83,87],
    'after':[85.5,89.9,92.6,89.5,85.8,88.8,84.6,87.8]
    })

#### 단측 검정 (before-after)의 평균 <0
- 계산 방식은 before-after이다.
- 대립 가설을 설명하려면 before가 더 작아야한다.
- 귀무가설 : 새로운 교육 프로그램은 효과가 없다. (μ >= 0)
- 대립가설 : 새로운 교육 프로그램은 효과가 있다. (μ < 0)

- ttest_rel() 함수를 사용해 대응 표본 검성 수행
- 대립가설에 따르면 새로운 교육 프로그램은 도입한 후의 성과가 이전보다 향상되었을 것이라고 예상
- alternative ='less' 옵션 사용


In [15]:
from scipy import stats
stats.ttest_rel(df['before'],df['after'], alternative='less')


# p-value가 0.05보다 작기 때문에 유의수준 0.05하에서 귀무가설 기각 -> 새로운 프로그램이 효과적

TtestResult(statistic=np.float64(-2.2127749675452324), pvalue=np.float64(0.03127028733756239), df=np.int64(7))

#### 단측 검정 (before-after)의 평균 > 0
- 계산 방식은 after-before이다.
- 대립 가설을 설명하려면 after 더 커야한다.
- 귀무가설 : 새로운 교육 프로그램은 효과가 없다. (μ <= 0)
- 대립가설 : 새로운 교육 프로그램은 효과가 있다. (μ > 0)

- ttest_rel() 함수를 사용해 대응 표본 검성 수행
- 대립가설에 따르면 새로운 교육 프로그램은 도입한 후의 성과가 이전보다 향상되었을 것이라고 예상


In [16]:
from scipy import stats
stats.ttest_rel(df['after'],df['before'], alternative='greater')


# p-value가 0.05보다 작기 때문에 유의수준 0.05하에서 귀무가설 기각 -> 새로운 프로그램이 효과적

TtestResult(statistic=np.float64(2.2127749675452324), pvalue=np.float64(0.03127028733756239), df=np.int64(7))

- (교육 전 - 교육 후) 평균
    - TtestResult(statistic=np.float64(-2.2127749675452324), pvalue=np.float64(0.03127028733756239), df=np.int64(7))
- (교육 후 - 교육 전) 평균
    - TtestResult(statistic=np.float64(2.2127749675452324), pvalue=np.float64(0.03127028733756239), df=np.int64(7))

#### 양측 검정
- μ = (교육 참여 전 성과 - 교육 참여 후 성과)의 평균
- 귀무가설 : 교육 전후 차이가 없음. (μ = 0)
- 대립가설 : 교육 전후 차이가 있다. (μ != 0)

In [17]:
stats.ttest_rel(df['after'],df['before'])

TtestResult(statistic=np.float64(2.2127749675452324), pvalue=np.float64(0.06254057467512478), df=np.int64(7))

#### 심화학습
- 문제 : 어떤 기업이 새로운 교육 프로그램을 도입해 직원의 성과를 개선하려고 한다.
- 이때 동일한 직원 그룹에 교육 전과 후의 성과를 비교했다.
- 새로운 교육 프로그램은 효과가 있는지 가설검정 하시오.(유의수준 0.05)
    - 귀무가설 : 새로운 교육 프로그램은 효과가 없다.
    - 대립가설 : 새로운 교육 프로그램은 효과가 있다.


In [18]:
import pandas as pd

df = pd.DataFrame({
    'before':[85,90,92,88,86,89,83,87],
    'after':[86,92,94,89,84,90,84,88]
})

- 샤피로-월크 검정은 주어진 샘플 데이터가 정규 분포를 따르는지의 여부를 검정하는 데 사용하며,
- 'after' , 'before' 변수의 순서는 결과에 영향을 주지 않는다.
```python
stats.shapiro()
```

In [19]:
from scipy import stats

df['diff'] = df['after']-df['before']
stats.shapiro(df['diff'])

ShapiroResult(statistic=np.float64(0.6886147375920879), pvalue=np.float64(0.0016734051223900109))

In [20]:
stats.wilcoxon(df['after'],df['before'],alternative='greater')

WilcoxonResult(statistic=np.float64(29.0), pvalue=np.float64(0.07421875))

## section 04 : 독립 표본 검정


### 독립 표본 검정
- 두 그룹 간의 평균이 서로 다름을 판단하는 통계방법이다.
- 예) 남자와 여자의 소득 차이가 있는지 없는지 가설 검정할 수 있다.
- a,b그룹으로 나누고 각 그룹을 대상으로 적절한 통계적 검정을 통해 두 그룹 사이에 통계적으로 
- 유의미한 차이가 있는지 없는지 결론을 내리는 것이 독립 표본 검정이다.

```python
ttest_ind(a,b,alternative, equal_val)
``` 
- a : 첫 번째 모집단에서 뽑은 표본데이터
- b : 두 번째 모집단에서 뽑은 표본데이터
- alternative : 대립가설 정의 : μ = (a-b)의 평균일 때
    - μ > 0 : greater , a의 평균이 b의 평균보다 크다.
    - μ < 0 : less , a의 평균이 b의 평균보다 작다.
    - μ != 0 : two-sided

- equal_val
    - True : 두 모집단의 분산이 같다고 가정
    - Fale : 두 모집단의 분산이 다르다고 가정

#### 문제 : 
- 어느 학교의 반별 수학 시험 점수이다. 1반과 2반의 평균 점수가 차이가 있는지
- 유의 수준 0.05하에서 가설 검정
- μ1 : 1반 평균 | μ2 : 2반 평균



##### 양측 검정
- 귀무가설 : 반별 수학 평균 점수는 같다
- 대립가설 : 반별 수학 평균 점수는 다르다

In [24]:
import pandas as pd
class1=[85,90,92,88,86,89,83,87]
class2=[80,82,88,85,84]


from scipy import stats

stats.ttest_ind(class1,class2, equal_var=False)
# equal_var는 기본값이 True이므로 모분산이 다르다면 false를 넣어줘야한다.

TtestResult(statistic=np.float64(2.1818699281825236), pvalue=np.float64(0.059589330071355334), df=np.float64(8.272682358753572))

##### 단측 검정 (- μ1 : 1반 평균 < μ2 : 2반 평균, 모분산 동일)
- 귀무가설 : 반별 수학 평균 점수는 같다
- 대립가설 : 2반 수학 평균 점수가 더 높다

In [25]:
stats.ttest_ind(class1,class2,equal_var=True,alternative='less')

TtestResult(statistic=np.float64(2.2108140580092237), pvalue=np.float64(0.9754257110537391), df=np.float64(11.0))

##### 단측 검정 (- μ1 : 1반 평균 > μ2 : 2반 평균, 모분산 동일)
- 귀무가설 : 반별 수학 평균 점수는 같다
- 대립가설 : 1반 수학 평균 점수가 더 높다.

In [26]:
stats.ttest_ind(class1,class2,equal_var=True,alternative='greater')

TtestResult(statistic=np.float64(2.2108140580092237), pvalue=np.float64(0.02457428894626093), df=np.float64(11.0))

#### 심화 문제
- 1반과 2반의 평균 점수가 차이가 있는지 유의 수준 0.05하에서 가설 검성
- 귀무가설 : 반별 수학 평균 점수는 같다. (μ1=μ2)
- 대립가설 : 2반 수학 평균 점수가 더 높다. (μ1<μ2)

In [28]:
# 모수 검정
import pandas as pd

class1=[85,90,92,88,86,89,83,87]
class2=[80,82,88,85,84]

from scipy import stats
print(stats.shapiro(class1))
print(stats.shapiro(class2))

# class1 p-value : 0.99
# class2 p-value : 0.98


ShapiroResult(statistic=np.float64(0.9981893537736595), pvalue=np.float64(0.999986994137081))
ShapiroResult(statistic=np.float64(0.9917398436295006), pvalue=np.float64(0.9854182266624975))


In [29]:
print(stats.levene(class1,class2))

# Levene(레빈) 검정을 수행한 결과, p-value가 0.95로 두 그룹은 등분산성을 만족


LeveneResult(statistic=np.float64(0.0027925869510027727), pvalue=np.float64(0.958802951766629))


In [30]:
# 독립 표본 t-검정 수행 
# 대립가설에 만족하려면 class1의 평균이 더 작아야 하므로
# alternative = 'less'로 설정한다
print(stats.ttest_ind(class1,class2, alternative='less',equal_var=True))

TtestResult(statistic=np.float64(2.2108140580092237), pvalue=np.float64(0.9754257110537391), df=np.float64(11.0))


In [31]:
# 비모수 검정
import pandas as pd

class1=[85,90,92,88,86,89,83,87]
class2=[80,82,88,85,130]

from scipy import stats
print(stats.shapiro(class1))
print(stats.shapiro(class2))

stats.mannwhitneyu(class1,class2,alternative='less')

ShapiroResult(statistic=np.float64(0.9981893537736595), pvalue=np.float64(0.999986994137081))
ShapiroResult(statistic=np.float64(0.6880497349322277), pvalue=np.float64(0.007151570728885509))


MannwhitneyuResult(statistic=np.float64(26.0), pvalue=np.float64(0.8299904236851448))